More ScalaTest-y specs

ducky64 opened the pull request:
Implementation for #73.

See the included test code for an API example.

Feedback is welcome on the more experimental ChiselImplicitPokeSpec. Especially if there is a way to properly overload triple-equals on FlatSpec.

chick commented on src/test/scala/chisel3/iotesters/ChiselPokeSpec.scala:
```diff
@@ -0,0 +1,40 @@
+// See LICENSE for license details.
```
Shouldn't this be ChiselPokeSpecSpec? :-)

ducky64 replied:
Good point, fixed!

chick commented on src/test/scala/chisel3/iotesters/ChiselPokeSpec.scala:
```diff
@@ -0,0 +1,40 @@
+class MyImplicitDutSpec extends ChiselImplicitPokeSpec {
+  "MyDut with implicits" should "properly add" in {
+    run(new MyDut) {implicit t => c =>
+      c.io.in <<= 0x41
```
I wonder if support for scalatest Matchers is worth considering to make this seem more like scalatest
```
peek(c.io.in) should be (0x0)
```

ducky64 replied:
I think that syntax is pretty verbose. Theoretically it could be done, though I don't see people preferring to write `should be` over `?==`.

chick commented on src/test/scala/chisel3/iotesters/ChiselPokeSpec.scala:
```diff
@@ -0,0 +1,40 @@
+  io.out := io.in + 1.U
+}
+
+class MyDutSpec extends ChiselPokeSpec {
```
It looks like ChiselPokeSpec is based on FlatSpec, I think you should consider FreeSpec
I think FreeSpec gives writer ability to write more expressive test harnesses

ducky64 replied:
Fixed, now you can choose your own *Spec type.

chick commented on src/test/scala/chisel3/iotesters/ChiselPokeSpec.scala:
```diff
@@ -0,0 +1,40 @@
+
+class MyDutSpec extends ChiselPokeSpec {
+  "MyDut" should "properly add" in {
+    run(new MyDut) {(t, c) =>
```
Just noting, that there was some resistance to ```run()``` approach.

ducky64 replied:
I don't particularly like this syntax, though the previous version where you had to manually synchronize the module type between your DUT, your spec, and your tester classes was completely insane. This at least keeps things very DRY and very concise.

chick commented:
I'd like to see some basic tools for managing decoupled interfaces, lack of such tools pushes users to invent their own solutions to this testing problem.  I think this hurts chisel adoption.

ducky64 commented:
Changed name to PokeTester/ImplicitPokeTester (I don't think we need Chisel_ on everything). Removed dependency on FlatSpec, so you can (probably) have a class that extends FreeSpec with ImplicitPokeTester.
Made ?== contain the assert inside, since ScalaTest doesn't understand assert(x ?== someVal) and doesn't print out helpful errors.

ducky64 commented:
@sdtwigg, thoughts on both style and implementation? (almost guaranteed Scala is being used wrong here somewhere)
@edwardcwang @jackkoenig

edwardcwang commented:
I'd be a little curious about why `<<=` becomes poke? At first sight it seems more natural that `x <<= y` means `x = x << y` or something.

Out of curiosity, why can't we just have the regular assignment operator for poke? e.g. let `poke(c.io.my_signal, 1)` be equivalent to just `c.io.my_signal := 1`? That way writing the testbench is like writing a module. Just a thought.

ducky64 commented:
In the short term, using := should be OK since nothing in Data uses that on Scala types, so it should look for an implicit conversion. However, I could also see cases in the future where we might want to pass in a Chisel literal, then since := is implemented on Data with Data, it will just resolve with the default semantics (generate a connection, which means nothing in test-land). I don't know of a way around this.

The approach here is just "use new operators so there's absolutely no ambiguity". The downside is that it introduces new operators, though my argument would be that it's just two, and these will get used over and over and over again.

kammoh commented:
can't DynamicContext.currentModule be checked to differentiate between RTL
and test code?

On Fri, Dec 16, 2016 at 8:06 AM Richard Lin <notifications@github.com>
wrote:

> In the short term, using := should be OK since nothing in Data uses that
> on Scala types, so it should look for an implicit conversion. However, I
> could also see cases in the future where we might want to pass in a Chisel
> literal, then since := is implemented on Data with Data, it will just
> resolve with the default semantics (generate a connection, which means
> nothing in test-land). I don't know of a way around this.
>
>
> The approach here is just "use new operators so there's absolutely no
> ambiguity". The downside is that it introduces new operators, though my
> argument would be that it's just two, and these will get used over and over
> and over again.
>
>
>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ucb-bar/chisel-testers/pull/74#issuecomment-267515222>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABWvAE7YU03gyW3j0pnS3KMIHlqzmRZiks5rIhU3gaJpZM4LOqf0>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

ducky64 commented:
Yes, but you would need infrastructure (in base Chisel) to allow those operators to do something different while not in hardware construction mode. That doesn't exist right now, and it's unclear whether that's a good idea.

ducky64 commented:
Any more comments on this? I really would like to get this merged if there's agreement that this is a good idea (structurally, overall) and that the details are good enough at least as a starting point for an experimental feature.

sdtwigg commented:
Why so many new operators? poke and expect easy to recall whereas <<= and ?== are mystical. Also, it is not visually evident that ?== is actually asserting (rather than doing an equality check).

For Boolean, we can define overloads that automatically do that conversion to Int.

ducky64 commented:
Resolution as follows:
- will remove the operators and change them back to poke and expect
- will change `run` to `test`

ducky64 commented:
So one problem with using `expect` is that it is already defined in scalaTest.Assertions, and the compiler can't seem to resolve to the right expect. That seems to be because an implicit conversion is needed from Int to BigInt, so it defaults back to the Assertions.expect, doesn't find enough arguments, and gives up.

One solution would be to change `expect` to something else, like `assertValue`.

Scala also has issues with multiple overloaded methods that define defaults, making it tedious to define versions of expect that take Booleans (now you have an expect for BigInt and an expect for Boolean, each version would then need to define a version with the explicit error message and one without, for a total of four - and eventually we would need to support FixedPoint / floats...)

ducky64 commented:
(or keep `expect` and don't rely on any kind of implicit conversions, which is compatible but limiting. A lighterweight (in number of characters) alternative is `check` which doesn't seem to be used either)

ducky64 commented:
Review please! (tests updated, it now uses `poke` and `check` because of the expect conflict mentioned above) @chick @jackkoenig

ducky64 commented:
**poke**
